Error mapping processor from Set to Array of Object

```java
package org.test;
import java.util.Set;
public class Entity {
    private Set<EntityExtra> extras;
    // getter and setter
}
```
```java
package org.test;
public class EntityExtra {
    private String value;
    // getter and setter
}
```
```java
package org.test;
public class Form {
    private FormExtra[] extras;
    // getter and setter
}
```
```java
package org.test;
public class FormExtra {
    private String value;
    // getter and setter
}
```
```java
package org.test;
import org.mapstruct.Mapper;
@Mapper
public interface FormMapper {
    Form from(Entity entity);
}
```
```xml
<plugin>
  <groupId>org.apache.maven.plugins</groupId>
  <artifactId>maven-compiler-plugin</artifactId>
  <version>3.8.0</version>
  <configuration>
    <source>1.8</source>
    <target>1.8</target>
    <annotationProcessorPaths>
      <path>
        <groupId>org.mapstruct</groupId>
        <artifactId>mapstruct-processor</artifactId>
        <version>1.3.1.Final</version>
      </path>
    </annotationProcessorPaths>
    <compilerArgs>
      <compilerArg>-Amapstruct.defaultComponentModel=spring</compilerArg>
    </compilerArgs>
  </configuration>
</plugin>
```

### Error building project:

```
package org.test;

FreeMarker template error:
An error has occurred when reading existing sub-variable "importTypeNames"; see cause exception! The type of the containing value was: extended_hash (org.mapstruct.ap.internal.model.Mapper wrapped into f.e.bded.freemarker.ext.beans.BeanModel)

----
FTL stack trace ("~" means nesting-related):
	- Failed at: #list importTypeNames as importedType  [in template "org/mapstruct/ap/internal/model/GeneratedType.ftl" at line 13, column 1]
----

Java stack trace (for programmers):
----
org.mapstruct.ap.shaded.freemarker.core._TemplateModelException: [... Exception message was already printed; see it above ...]
	at org.mapstruct.ap.shaded.freemarker.ext.beans.BeanModel.get(BeanModel.java:198)
	at org.mapstruct.ap.internal.writer.FreeMarkerModelElementWriter$ExternalParamsTemplateModel.get(FreeMarkerModelElementWriter.java:60)
	at org.mapstruct.ap.shaded.freemarker.core.Environment.getGlobalVariable(Environment.java:1403)
	at org.mapstruct.ap.shaded.freemarker.core.Environment.getVariable(Environment.java:1389)
	at org.mapstruct.ap.shaded.freemarker.core.Identifier._eval(Identifier.java:35)
	at org.mapstruct.ap.shaded.freemarker.core.Expression.eval(Expression.java:76)
	at org.mapstruct.ap.shaded.freemarker.core.IteratorBlock.accept(IteratorBlock.java:58)
	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)
	at org.mapstruct.ap.shaded.freemarker.core.MixedContent.accept(MixedContent.java:57)
	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)
	at org.mapstruct.ap.shaded.freemarker.core.Environment.process(Environment.java:235)
	at org.mapstruct.ap.shaded.freemarker.template.Template.process(Template.java:262)
	at org.mapstruct.ap.internal.writer.FreeMarkerModelElementWriter.write(FreeMarkerModelElementWriter.java:35)
	at org.mapstruct.ap.internal.writer.FreeMarkerWritable.write(FreeMarkerWritable.java:19)
	at org.mapstruct.ap.internal.writer.ModelWriter.writeModel(ModelWriter.java:71)
	at org.mapstruct.ap.internal.processor.MapperRenderingProcessor.createSourceFile(MapperRenderingProcessor.java:62)
	at org.mapstruct.ap.internal.processor.MapperRenderingProcessor.writeToSourceFile(MapperRenderingProcessor.java:39)
	at org.mapstruct.ap.internal.processor.MapperRenderingProcessor.process(MapperRenderingProcessor.java:29)
	at org.mapstruct.ap.internal.processor.MapperRenderingProcessor.process(MapperRenderingProcessor.java:24)
	at org.mapstruct.ap.MappingProcessor.process(MappingProcessor.java:283)
	at org.mapstruct.ap.MappingProcessor.processMapperTypeElement(MappingProcessor.java:263)
	at org.mapstruct.ap.MappingProcessor.processMapperElements(MappingProcessor.java:221)
	at org.mapstruct.ap.MappingProcessor.process(MappingProcessor.java:156)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:794)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:705)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1800(JavacProcessingEnvironment.java:91)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1035)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1176)
	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)
	at com.sun.tools.javac.main.Main.compile(Main.java:523)
	at com.sun.tools.javac.main.Main.compile(Main.java:381)
	at com.sun.tools.javac.main.Main.compile(Main.java:370)
	at com.sun.tools.javac.main.Main.compile(Main.java:361)
	at com.sun.tools.javac.Main.compile(Main.java:56)
	at com.sun.tools.javac.Main.main(Main.java:42)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.mapstruct.ap.shaded.freemarker.ext.beans.BeansWrapper.invokeMethod(BeansWrapper.java:1385)
	at org.mapstruct.ap.shaded.freemarker.ext.beans.BeanModel.invokeThroughDescriptor(BeanModel.java:254)
	at org.mapstruct.ap.shaded.freemarker.ext.beans.BeanModel.get(BeanModel.java:158)
	... 35 more
Caused by: java.lang.NullPointerException
	at org.mapstruct.ap.internal.model.GeneratedType.needsImportDeclaration(GeneratedType.java:291)
	at org.mapstruct.ap.internal.model.GeneratedType.addIfImportRequired(GeneratedType.java:269)
	at org.mapstruct.ap.internal.model.GeneratedType.getImportTypes(GeneratedType.java:219)
	at org.mapstruct.ap.internal.model.GeneratedType.getImportTypeNames(GeneratedType.java:250)
	... 42 more
```

--- a/processor/src/main/java/org/mapstruct/ap/internal/model/GeneratedType.java
+++ b/processor/src/main/java/org/mapstruct/ap/internal/model/GeneratedType.java
@@ -288,8 +288,15 @@ public abstract class GeneratedType extends ModelElement {
             }
 
             if ( typeToAdd.getPackageName().equals( packageName ) ) {
-                if ( !typeToAdd.getTypeElement().getNestingKind().isNested() ) {
-                    return false;
+                if ( typeToAdd.getTypeElement() != null ) {
+                    if ( !typeToAdd.getTypeElement().getNestingKind().isNested() ) {
+                        return false;
+                    }
+                }
+                else if ( typeToAdd.getComponentType() != null ) {
+                    if ( !typeToAdd.getComponentType().getTypeElement().getNestingKind().isNested() ) {
+                        return false;
+                    }
                 }
             }
         }

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2001/Entity.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2001/Entity.java
@@ -1,0 +1,25 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2001;
+
+import java.util.Set;
+
+/**
+ * @author Filip Hrisafov
+ */
+public
+class Entity {
+
+    private Set<EntityExtra> extras;
+
+    public Set<EntityExtra> getExtras() {
+        return extras;
+    }
+
+    public void setExtras(Set<EntityExtra> extras) {
+        this.extras = extras;
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2001/EntityExtra.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2001/EntityExtra.java
@@ -1,0 +1,23 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2001;
+
+/**
+ * @author Filip Hrisafov
+ */
+public
+class EntityExtra {
+
+    private String value;
+
+    public String getValue() {
+        return value;
+    }
+
+    public void setValue(String value) {
+        this.value = value;
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2001/Form.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2001/Form.java
@@ -1,0 +1,23 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2001;
+
+/**
+ * @author Filip Hrisafov
+ */
+public
+class Form {
+
+    private FormExtra[] extras;
+
+    public FormExtra[] getExtras() {
+        return extras;
+    }
+
+    public void setExtras(FormExtra[] extras) {
+        this.extras = extras;
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2001/FormExtra.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2001/FormExtra.java
@@ -1,0 +1,23 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2001;
+
+/**
+ * @author Filip Hrisafov
+ */
+public
+class FormExtra {
+
+    private String value;
+
+    public String getValue() {
+        return value;
+    }
+
+    public void setValue(String value) {
+        this.value = value;
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2001/Issue2001Mapper.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2001/Issue2001Mapper.java
@@ -1,0 +1,17 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2001;
+
+import org.mapstruct.Mapper;
+
+/**
+ * @author Filip Hrisafov
+ */
+@Mapper
+public interface Issue2001Mapper {
+
+    Form map(Entity entity);
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2001/Issue2001Test.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2001/Issue2001Test.java
@@ -1,0 +1,32 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2001;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mapstruct.ap.testutil.IssueKey;
+import org.mapstruct.ap.testutil.WithClasses;
+import org.mapstruct.ap.testutil.runner.AnnotationProcessorTestRunner;
+
+/**
+ * @author Filip Hrisafov
+ */
+@IssueKey("2001")
+@RunWith( AnnotationProcessorTestRunner.class )
+@WithClasses( {
+    Entity.class,
+    EntityExtra.class,
+    Form.class,
+    FormExtra.class,
+    Issue2001Mapper.class
+} )
+public class Issue2001Test {
+
+    @Test
+    public void shouldCompile() {
+
+    }
+}

